constructors.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+type scopedProvider struct {
+	Provider
+
+	prefix string
+}
+
+var _ Provider = (*scopedProvider)(nil)
+
+func (s *scopedProvider) Get(key string) Value {
+	return s.Provider.Get(fmt.Sprintf("%s.%s", s.prefix, key))
+}
+
+// NewScopedProvider wraps a provider and adds a prefix to all Get calls.
+func NewScopedProvider(prefix string, provider Provider) Provider {
+	if prefix == "" {
+		return provider
+	}
+	return &scopedProvider{provider, prefix}
+}
+
+// NewProviderGroup composes multiple providers, with later providers
+// overriding earlier ones. The merge logic is described in the package-level
+// documentation. To preserve backward compatibility, the resulting provider
+// disables strict unmarshalling.
+//
+// Deprecated: compose multiple sources of configuration using NewYAML and the
+// Static, File, and/or Source options directly. This enables strict
+// unmarshalling by default and allows use of other options at the same time.
+func NewProviderGroup(name string, providers ...Provider) (Provider, error) {
+	opts := make([]YAMLOption, 0, len(providers)+2)
+	opts = append(opts, Name(name), Permissive())
```

Comment:
Ah, I see the use of the `Name` option now. Feel free to ignore my other comment.